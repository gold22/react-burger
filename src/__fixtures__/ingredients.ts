import { TIngredient } from '../services/types';

// eslint-disable-next-line import/prefer-default-export
export const ingredient: TIngredient = {
    id: '60d3b41abdacab0026a733c6',
    name: 'Краторная булка N-200i',
    type: 'bun',
    proteins: 80,
    fat: 24,
    carbohydrates: 53,
    calories: 420,
    price: 1255,
    image: 'https://code.s3.yandex.net/react/code/bun-02.png',
    imageLarge: 'https://code.s3.yandex.net/react/code/bun-02-large.png',
    imageMobile: 'https://code.s3.yandex.net/react/code/bun-02-mobile.png',
};
