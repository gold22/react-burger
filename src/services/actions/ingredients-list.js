export const GET_INGREDIENTS_REQUEST = 'GET_INGREDIENTS_REQUEST';
export const GET_INGREDIENTS_SUCCESS = 'GET_INGREDIENTS_SUCCESS';
export const GET_INGREDIENTS_FAILED = 'GET_INGREDIENTS_FAILED';

export const getIngredients = (apiClient) => {
    return async function (dispatch) {
        dispatch({
            type: GET_INGREDIENTS_REQUEST,
        });
        try {
            const ingredients = await apiClient.getIngredients();
            dispatch({
                type: GET_INGREDIENTS_SUCCESS,
                ingredients,
            });
        } catch (error) {
            dispatch({
                type: GET_INGREDIENTS_FAILED,
                message: error.message,
            });
        }
    };
}
