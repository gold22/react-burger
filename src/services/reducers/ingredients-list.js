import {
    GET_INGREDIENTS_REQUEST,
    GET_INGREDIENTS_SUCCESS,
    GET_INGREDIENTS_ERROR,
} from '../actions/ingredients-list';

const initialState = {
    ingredients: [],
    isLoading: true,
    loadError: null,
};

// eslint-disable-next-line import/prefer-default-export
export const ingredientsListReducer = (state = initialState, action) => {
    switch (action.type) {
    case GET_INGREDIENTS_REQUEST: {
        return {
            ...state,
            isLoading: true,
        };
    }
    case GET_INGREDIENTS_SUCCESS: {
        return {
            ...state,
            isLoading: false,
            loadError: null,
            ingredients: action.ingredients,
        };
    }
    case GET_INGREDIENTS_ERROR: {
        return {
            ...initialState,
            isLoading: false,
            loadError: action.message,
        };
    }
    default: {
        return state;
    }
    }
};
