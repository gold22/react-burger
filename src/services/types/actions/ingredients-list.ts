import {
    GET_INGREDIENTS_REQUEST,
    GET_INGREDIENTS_SUCCESS,
    GET_INGREDIENTS_ERROR,
} from '../../constants/ingredients-list';
import { TIngredients } from '../index';

export interface IGetIngredientsAction {
    readonly type: typeof GET_INGREDIENTS_REQUEST;
}

export interface IGetIngredientsSuccessAction {
    readonly type: typeof GET_INGREDIENTS_SUCCESS;
    readonly ingredients: TIngredients;
}

export interface IGetIngredientsFailedAction {
    readonly type: typeof GET_INGREDIENTS_ERROR;
    readonly message: string;
}

export type TIngredientsListActions =
    | IGetIngredientsAction
    | IGetIngredientsSuccessAction
    | IGetIngredientsFailedAction;
