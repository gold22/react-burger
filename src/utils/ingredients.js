export const INGREDIENT_TYPE_BUN = 'bun';
export const INGREDIENT_TYPE_SAUCE = 'sauce';
export const INGREDIENT_TYPE_MAIN = 'main';

export const getIngredientCount = (ingredientId, ingredients) => (
    ingredients.reduce(
        (count, ingredient) => (ingredientId === ingredient.id ? count + 1 : count),
        0,
    )
);

export const getIngredientsPrice = (ingredients) => (
    ingredients.reduce((sum, ingredient) => sum + ingredient.price, 0)
);

export const getIngredientsIds = (ingredients) => (
    ingredients.map((ingredient) => ingredient.id)
);

export const isBun = (ingredient) => (
    ingredient.type === INGREDIENT_TYPE_BUN
);

export const isSauce = (ingredient) => (
    ingredient.type === INGREDIENT_TYPE_SAUCE
);

export const isMain = (ingredient) => (
    ingredient.type === INGREDIENT_TYPE_MAIN
);

export const getBun = (ingredients) => (
    ingredients.find((ingredient) => isBun(ingredient))
);
