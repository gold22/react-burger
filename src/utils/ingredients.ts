import { TIngredient, TIngredients } from './ts-types';

export const INGREDIENT_TYPE_BUN = 'bun';
export const INGREDIENT_TYPE_SAUCE = 'sauce';
export const INGREDIENT_TYPE_MAIN = 'main';

export const getIngredient = (ingredientId: string, ingredients: TIngredients) => (
    ingredients.find((ingredient) => ingredientId === ingredient.id)
);

export const getIngredientCount = (ingredientId: string, ingredients: TIngredients) => (
    ingredients.reduce(
        (count, ingredient) => (ingredientId === ingredient.id ? count + 1 : count),
        0,
    )
);

export const getIngredientsPrice = (ingredients: TIngredients) => (
    ingredients.reduce((sum, ingredient) => sum + ingredient.price, 0)
);

export const getIngredientsIds = (ingredients: TIngredients) => (
    ingredients.map((ingredient) => ingredient.id)
);

export const isBun = (ingredient: TIngredient) => (
    ingredient.type === INGREDIENT_TYPE_BUN
);

export const isSauce = (ingredient: TIngredient) => (
    ingredient.type === INGREDIENT_TYPE_SAUCE
);

export const isMain = (ingredient: TIngredient) => (
    ingredient.type === INGREDIENT_TYPE_MAIN
);

export const getBun = (ingredients: TIngredients) => (
    ingredients.find((ingredient) => isBun(ingredient))
);
