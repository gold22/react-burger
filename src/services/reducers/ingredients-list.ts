import {
    GET_INGREDIENTS_REQUEST,
    GET_INGREDIENTS_SUCCESS,
    GET_INGREDIENTS_ERROR,
} from '../constants/ingredients-list';
import { TIngredients } from '../types';
import { TIngredientsListActions } from '../types/actions/ingredients-list';

export type TIngredientsListState = {
    ingredients: TIngredients;
    isLoading: boolean;
    loadError: string | null;
};

const initialState: TIngredientsListState = {
    ingredients: [],
    isLoading: true,
    loadError: null,
};

export const ingredientsListReducer = (
    // eslint-disable-next-line @typescript-eslint/default-param-last
    state: TIngredientsListState = initialState,
    action: TIngredientsListActions,
): TIngredientsListState => {
    switch (action.type) {
    case GET_INGREDIENTS_REQUEST: {
        return {
            ...state,
            isLoading: true,
        };
    }
    case GET_INGREDIENTS_SUCCESS: {
        return {
            ...state,
            isLoading: false,
            loadError: null,
            ingredients: action.ingredients,
        };
    }
    case GET_INGREDIENTS_ERROR: {
        return {
            ...initialState,
            isLoading: false,
            loadError: action.message,
        };
    }
    default: {
        return state;
    }
    }
};
