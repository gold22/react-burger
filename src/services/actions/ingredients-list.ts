import {
    GET_INGREDIENTS_REQUEST,
    GET_INGREDIENTS_SUCCESS,
    GET_INGREDIENTS_ERROR,
} from '../constants/ingredients-list';
import { TDispatch, TGetState, TThunk } from '../types/store';
import ApiClient from '../api-client';
import { getErrorMessage } from '../../utils/error';

// eslint-disable-next-line import/prefer-default-export
export const getIngredients: TThunk = () => async (
    dispatch: TDispatch,
    getState: TGetState,
    apiClient: ApiClient,
) => {
    dispatch({
        type: GET_INGREDIENTS_REQUEST,
    });
    try {
        const ingredients = await apiClient.getIngredients();
        dispatch({
            type: GET_INGREDIENTS_SUCCESS,
            ingredients: ingredients.map((ingredient) => ({
                ...ingredient,
                // eslint-disable-next-line no-underscore-dangle
                id: ingredient._id,
                imageLarge: ingredient.image_large,
                imageMobile: ingredient.image_mobile,
            })),
        });
    } catch (error) {
        dispatch({
            type: GET_INGREDIENTS_ERROR,
            message: getErrorMessage(error),
        });
    }
};
