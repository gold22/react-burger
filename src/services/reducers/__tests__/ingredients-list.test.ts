import { ingredientsListReducer, TIngredientsListState } from '../ingredients-list';
import {
    GET_INGREDIENTS_ERROR,
    GET_INGREDIENTS_REQUEST,
    GET_INGREDIENTS_SUCCESS,
} from '../../constants/ingredients-list';
import { ingredients } from '../../../__fixtures__/ingredients';

const initialState: TIngredientsListState = ingredientsListReducer(undefined, {} as any);

describe('ingredients list reducer', () => {
    it('should return the initial state', () => {
        expect(initialState).toMatchSnapshot();
    });

    it('should handle GET_INGREDIENTS_REQUEST', () => {
        expect(ingredientsListReducer(undefined, { type: GET_INGREDIENTS_REQUEST }))
            .toMatchSnapshot();
    });

    it('should handle GET_INGREDIENTS_SUCCESS', () => {
        const state = {
            ...initialState,
            isLoading: true,
        };
        expect(ingredientsListReducer(state, { type: GET_INGREDIENTS_SUCCESS, ingredients }))
            .toMatchSnapshot();
    });

    it('should handle GET_INGREDIENTS_ERROR', () => {
        const state = {
            ...initialState,
            isLoading: true,
        };
        expect(ingredientsListReducer(state, { type: GET_INGREDIENTS_ERROR, message: 'Get ingredients error' }))
            .toMatchSnapshot();
    });
});
