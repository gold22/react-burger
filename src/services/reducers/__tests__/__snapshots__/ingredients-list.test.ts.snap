// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ingredients list reducer should handle GET_INGREDIENTS_ERROR 1`] = `
Object {
  "ingredients": Array [],
  "isLoading": false,
  "loadError": "Get ingredients error",
}
`;

exports[`ingredients list reducer should handle GET_INGREDIENTS_REQUEST 1`] = `
Object {
  "ingredients": Array [],
  "isLoading": true,
  "loadError": null,
}
`;

exports[`ingredients list reducer should handle GET_INGREDIENTS_SUCCESS 1`] = `
Object {
  "ingredients": Array [
    Object {
      "calories": 420,
      "carbohydrates": 53,
      "fat": 24,
      "id": "60d3b41abdacab0026a733c6",
      "image": "https://code.s3.yandex.net/react/code/bun-02.png",
      "imageLarge": "https://code.s3.yandex.net/react/code/bun-02-large.png",
      "imageMobile": "https://code.s3.yandex.net/react/code/bun-02-mobile.png",
      "name": "Краторная булка N-200i",
      "price": 1255,
      "proteins": 80,
      "type": "bun",
    },
    Object {
      "calories": 643,
      "carbohydrates": 85,
      "fat": 26,
      "id": "60d3b41abdacab0026a733c7",
      "image": "https://code.s3.yandex.net/react/code/bun-01.png",
      "imageLarge": "https://code.s3.yandex.net/react/code/bun-01-large.png",
      "imageMobile": "https://code.s3.yandex.net/react/code/bun-01-mobile.png",
      "name": "Флюоресцентная булка R2-D3",
      "price": 988,
      "proteins": 44,
      "type": "bun",
    },
    Object {
      "calories": 643,
      "carbohydrates": 85,
      "fat": 26,
      "id": "60d3b41abdacab0026a733c8",
      "image": "https://code.s3.yandex.net/react/code/meat-03.png",
      "imageLarge": "https://code.s3.yandex.net/react/code/meat-03-large.png",
      "imageMobile": "https://code.s3.yandex.net/react/code/meat-03-mobile.png",
      "name": "Филе Люминесцентного тетраодонтимформа",
      "price": 988,
      "proteins": 44,
      "type": "main",
    },
    Object {
      "calories": 30,
      "carbohydrates": 40,
      "fat": 20,
      "id": "60d3b41abdacab0026a733cc",
      "image": "https://code.s3.yandex.net/react/code/sauce-02.png",
      "imageLarge": "https://code.s3.yandex.net/react/code/sauce-02-large.png",
      "imageMobile": "https://code.s3.yandex.net/react/code/sauce-02-mobile.png",
      "name": "Соус Spicy-X",
      "price": 90,
      "proteins": 30,
      "type": "sauce",
    },
  ],
  "isLoading": false,
  "loadError": null,
}
`;

exports[`ingredients list reducer should return the initial state 1`] = `
Object {
  "ingredients": Array [],
  "isLoading": true,
  "loadError": null,
}
`;
